### PR TITLE
add support for generice fallback method with just execption as a param plus some code refactoring

### PR DESCRIPTION
**Pull request by Romeh**

### CODE REVIEW

File: resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019 Kyuhyen Hwang
+ * Copyright 2019 Kyuhyen Hwang, Mahmoud Romih
```
Comment (RobWin) [inline resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java]: Typo in your name ;)

File: resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java [hunk 2]
```diff
@@ -15,28 +15,29 @@
...
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StringUtils;
+
+import io.github.resilience4j.core.lang.Nullable;
+
 /**
- * Reflection utility for invoking a fallback method. A fallback method should have the same return type and parameter types of original method but the last additional parameter.
- * The last additional parameter should be a subclass of {@link Throwable}. When {@link FallbackMethod#recover(Throwable)} is invoked, {@link Throwable} will be passed to that last parameter.
- * If there are multiple fallback methods, one of the methods that has most closest superclass parameter of thrown object will be invoked.
+ * Reflection utility for invoking fallbackMethod method. Recovery method should have same return type and parameter types of original method but the last additional parameter.
```
Comment (RobWin) [inline resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java]: Renaming has also overwritten this javadoc

File: resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java [hunk 3]
```diff
@@ -111,31 +110,111 @@ public Object recover(Throwable thrown) throws Throwable {
...
+        private String recoveryMethodName;
+        private Method originalMethod;
+        private Object[] args;
+        private Object target;
+
+        /**
+         * @param recoveryMethodName the configured recovery method name
+         * @return Builder of FallbackMethod
+         */
+        public Builder recoveryMethodName(String recoveryMethodName) {
```
Comment (RobWin) [inline resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java]: Should be fallbackMethodName

File: resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java [hunk 4]
```diff
@@ -111,31 +110,111 @@ public Object recover(Throwable thrown) throws Throwable {
...
+                fallback.setAccessible(false);
+            }
+        }
+    }
+
+    /**
+     * the FallbackMethod builder class
+     */
+    public static class Builder {
+        private String recoveryMethodName;
```
Comment (RobWin) [inline resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackMethod.java]: fallbackMethodName

File: resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerAspect.java [hunk 5]
```diff
@@ -97,7 +98,10 @@ public Object circuitBreakerAroundAdvice(ProceedingJoinPoint proceedingJoinPoint
 			return proceed(proceedingJoinPoint, methodName, circuitBreaker, returnType);
 		}
 
-		FallbackMethod fallbackMethod = new FallbackMethod(circuitBreakerAnnotation.fallbackMethod(), method, proceedingJoinPoint.getArgs(), proceedingJoinPoint.getTarget());
+		FallbackMethod fallbackMethod = FallbackMethod.builder().recoveryMethodName(circuitBreakerAnnotation.fallbackMethod())
```
Comment (RobWin) [inline resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerAspect.java]: Is anything optional? I think every parameter (recoveryMethodName, originalMethod, targetObject) is mandatory or?
The builder pattern should be used if some parameters are optional.

Comment (Romeh) [reply resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerAspect.java]: The building of Fallback method is not straight forward and i did not like that we have logic in the constructor , but u are right with the optional params with the builder pattern , i replaced it with simple static factory method now

**Pull request merged**